src/main/scala/io/picnicml/doddlemodel/preprocessing/QuantileDiscretizer.scala

```diff
@@ -0,0 +1,101 @@
+package io.picnicml.doddlemodel.preprocessing
+
+import breeze.linalg.DenseVector
+import breeze.stats.DescriptiveStats
+import io.picnicml.doddlemodel.CrossScalaCompat._
+import io.picnicml.doddlemodel.data.Feature.FeatureIndex
+import io.picnicml.doddlemodel.data.Features
+import io.picnicml.doddlemodel.syntax.OptionSyntax._
+import io.picnicml.doddlemodel.typeclasses.Transformer
+import scala.Double.{MinValue, MaxValue}
+
+case class QuantileDiscretizer(private val bucketCounts: DenseVector[Double], private val featureIndex: FeatureIndex, private val quantiles: Option[Seq[Seq[(Double, Double)]]] = None) {
+  private val numNumeric = featureIndex.numerical.columnIndices.length
+  require(numNumeric == 0 || numNumeric == bucketCounts.length, "A quantile should be given for every numerical column")
+}
+
+/** An immutable preprocessor that discretizes numerical features into buckets based on quantiles.
+  * Numerical feature values are converted to categorical features based on their distribution
+  * */
+object QuantileDiscretizer {
+
+  /** Create a quantile discretizer which splits data into discrete evenly sized buckets.
+    *
+    * @param bucketCounts The number of quantiles desired
+    * @param featureIndex feature index associated with features - this is needed so that only numerical features are
+    *                     transformed by this preprocessor; could be a subset of columns to be transformed
+    *
+    * @example Discretize a matrix into quartiles with two features: one numerical and one categorical.
+    *   {{{
+    *     import io.picnicml.doddlemodel.preprocessing.QuantileDiscretizer.ev
+    *
+    *     val featureIndex = FeatureIndex(List(NumericalFeature, CategoricalFeature))
+    *     val x = DenseMatrix(
+    *       List(-1.0, 0.0),
+    *       List(0.0, 1.0),
+    *       List(2.0, 0.0),
+    *       List(5.0, 0.0)
+    *     )
+    *     // equivalently, DenseVector(4) could be used
+    *     val bucketCounts = 4
+    *     val discretizer = QuantileDiscretizer(bucketCounts, featureIndex)
+    *     // Note: no fitting required
```

Comment:
This should be removed as the new version requires calling `fit`.

Reply:
Yup!